sdk/core/core-client/src/utils.ts

```diff
@@ -82,30 +146,34 @@ export function flattenResponse(
           arrayResponse[key] = parsedHeaders[key];
         }
       }
-      return arrayResponse;
+      return !fullResponse.parsedBody && isNullable && arrayResponse === emptyArray
+        ? null
+        : arrayResponse;
     }
 
     if (typeName === "Composite" || typeName === "Dictionary") {
-      return {
-        ...parsedHeaders,
-        ...fullResponse.parsedBody
-      };
+      return handleNullableResponseAndWrappableBody({
+        body: fullResponse.parsedBody,
+        headers: parsedHeaders,
+        hasNullableType: isNullable,
+        shouldWrapBody: false
+      });
     }
   }
 
-  if (
-    bodyMapper ||
-    fullResponse.request.method === "HEAD" ||
-    isPrimitiveType(fullResponse.parsedBody)
-  ) {
-    return {
-      ...parsedHeaders,
-      body: fullResponse.parsedBody
-    };
+  if (isPrimitiveType(fullResponse.parsedBody)) {
+    return handleNullableResponseAndWrappableBody({
+      body: fullResponse.parsedBody,
+      headers: parsedHeaders,
+      hasNullableType: isNullable,
+      shouldWrapBody: true
+    });
```

Comment:
Can't we ditch the if statement here and write something like this?

```suggestion
    return handleNullableResponseAndWrappableBody({
      body: fullResponse.parsedBody,
      headers: parsedHeaders,
      hasNullableType: isNullable,
      shouldWrapBody: isPrimitiveType(fullResponse.parsedBody)
    });
```